hadoop-hdds/rocks-native/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedSSTDumpIterator.java

```diff
@@ -0,0 +1,238 @@
+                                long pollIntervalMillis) throws IOException,
+          NativeLibraryNotLoadedException {
+    File sstFile = new File(sstFilePath);
+    if (!sstFile.exists() || !sstFile.isFile()) {
+      throw new IOException(String.format("Invalid SST File Path : %s",
+              sstFile.getAbsolutePath()));
+    }
+    this.pollIntervalMillis = pollIntervalMillis;
+    this.lock = new ReentrantLock();
+    init(sstDumpTool, sstFile, options);
+  }
+
+  private void init(ManagedSSTDumpTool sstDumpTool, File sstFile,
+                    ManagedOptions options)
+          throws NativeLibraryNotLoadedException {
+    String[] args = {"--file=" + sstFile.getAbsolutePath(),
+                     "--command=scan"};
+    this.sstDumpToolTask = sstDumpTool.run(args, options);
+    processOutput = new BufferedReader(new InputStreamReader(
+            sstDumpToolTask.getPipedOutput(), StandardCharsets.UTF_8));
+    stdoutString = new StringBuilder();
+    currentMatcher = PATTERN_MATCHER.matcher(stdoutString);
+    charBuffer = new char[8192];
+    open = new AtomicBoolean(true);
+    next();
+  }
+
+  /**
+   * Throws Runtime exception in the case iterator is closed or
+   * the native Dumptool exited with non zero exit value.
+   */
+  private void checkSanityOfProcess() {
+    if (!this.open.get()) {
+      throw new RuntimeException("Iterator has been closed");
+    }
+    if (sstDumpToolTask.getFuture().isDone()
+            && sstDumpToolTask.exitValue() != 0) {
+      throw new RuntimeException("Process Terminated with non zero " +
+              String.format("exit value %d", sstDumpToolTask.exitValue()));
+    }
+  }
+
+  /**
+   *
+   * @return
+   * Throws Runtime Exception in case of SST File read failure
+   */
+
+  @Override
+  public boolean hasNext() {
+    checkSanityOfProcess();
+    return nextKey != null;
+  }
+
+  /**
+   *
+   * @return next Key
+   * Throws Runtime Exception incase of failure.
+   */
+  @Override
+  public KeyValue next() {
+    lock.lock();
+    try {
+      checkSanityOfProcess();
+      currentKey = nextKey;
+      nextKey = null;
+      while (!currentMatcher.find()) {
+        try {
+          if (prevMatchEndIndex != 0) {
+            stdoutString = new StringBuilder(stdoutString.substring(
+                    prevMatchEndIndex, stdoutString.length()));
+            prevMatchEndIndex = 0;
+            currentMatcher = PATTERN_MATCHER.matcher(stdoutString);
+          }
+          Thread.sleep(pollIntervalMillis);
+          int numberOfCharsRead = processOutput.read(charBuffer);
+          if (numberOfCharsRead < 0) {
+            if (currentKey != null) {
+              currentKey.setValue(stdoutString.toString());
+            }
+            return currentKey;
+          }
+          stdoutString.append(charBuffer, 0, numberOfCharsRead);
+          currentMatcher.reset();
+        } catch (IOException | InterruptedException e) {
+          throw new RuntimeIOException(e);
```

Comment:
Consider if just throwing a checked exception is better

Reply:
Cannot throw exception since this implements Iterator